### PR TITLE
ConfigDrive fixes: CLOUDSTACK-10288, CLOUDSTACK-10289

### PR DESCRIPTION
**Pull request by fmaximus**
## Description
Bugfixes
- CLOUDSTACK-10288: Config Drive Userdata: support for binary userdata 
   Moved base64 decoding to ISO generation
- CLOUDSTACK-10289: Config Drive Metadata: Use VM UUID instead of VM id 
  Aligned Config Drive metadata code with VR metadata code.
- CLOUDSTACK-10358: Config Drive SSH Keys missing in come cases
  Taking SSH keys from VM details instead of vmProfile Parameter

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
Environment: KVM + Nuage
Tested manually, and executed the adapted marvin test.

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [x] I have added tests to cover my changes.
- [x] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

File: services/secondary-storage/server/src/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 1]
```diff
@@ -480,7 +480,7 @@ public Answer createConfigDriveIsoForVM(HandleConfigDriveIsoCommand cmd) {
                 for (String[] item : cmd.getVmData()) {
                     String dataType = item[CONFIGDATA_DIR];
                     String fileName = item[CONFIGDATA_FILE];
-                    String content = item[CONFIGDATA_CONTENT];
+                    String content = item[CONFIGDATA_CONTENT]; // base64
```
Comment (rhtyd) [inline services/secondary-storage/server/src/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java]: you can rename the variable to `base64Content`

Comment (DaanHoogland) [reply services/secondary-storage/server/src/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java]: and remove the comment?

File: test/integration/plugins/nuagevsp/nuage_lib.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+from marvin.cloudstackAPI import createSSHKeyPair, deleteSSHKeyPair
```
Comment (DaanHoogland) [inline test/integration/plugins/nuagevsp/nuage_lib.py]: apache license is missing

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@rhtyd @rafaelweingartner can you approve of this? I would like to base some new work on it ;)

**Comment by DaanHoogland** [comment]
@fmaximus I think I found a bug. I will retest but here is the issue:
when creating an offering with all options on virtual router except for user data (on configdrive) and creating a template with password enabled it creates a vm in a new network of this new offering type but does not supply the generated password. See the screenshot to see the issue.
screenshot from 2018-04-19 09-15-31

**Comment by fmaximus** [comment]
I found, and fixed another bug, CLOUDSTACK-10358, about SSH keypair not ending up on the configdrive disk in some cases.

File: test/integration/plugins/nuagevsp/nuage_lib.py (hunk 2 above)
Comment (fmaximus) [reply test/integration/plugins/nuagevsp/nuage_lib.py]: Fixed in commit 124d091

File: server/test/com/cloud/network/element/ConfigDriveNetworkElementTest.java [hunk 3]
```diff
@@ -236,13 +240,79 @@ public void testGetCapabilities () {
...
+
+    private List<String[]> getVmData() throws InsufficientCapacityException, ResourceUnavailableException {
         final Answer answer = mock(Answer.class);
         final UserVmDetailVO userVmDetailVO = mock(UserVmDetailVO.class);
         when(endpoint.sendMessage(any(HandleConfigDriveIsoCommand.class))).thenReturn(answer);
         when(answer.getResult()).thenReturn(true);
         when(network.getTrafficType()).thenReturn(Networks.TrafficType.Guest);
         when(virtualMachine.getState()).thenReturn(VirtualMachine.State.Stopped);
         when(userVmDetailVO.getValue()).thenReturn(PUBLIC_KEY);
+        when(nicp.getIPv4Address()).thenReturn("192.168.111.111");
```
Comment (nvazquez) [inline server/test/com/cloud/network/element/ConfigDriveNetworkElementTest.java]: Why not this one also as a final variable to use on the assert methods?

File: server/src/com/cloud/network/NetworkModelImpl.java [hunk 4]
```diff
@@ -2344,18 +2345,47 @@ public boolean getNetworkEgressDefaultPolicy(Long networkId) {
...
+
+        if (vmUuid == null) {
+            vmData.add(new String[]{METATDATA_DIR, INSTANCE_ID_FILE, vmName});
+            vmData.add(new String[]{METATDATA_DIR, VM_ID_FILE, String.valueOf(vmId)});
+        } else {
+            vmData.add(new String[]{METATDATA_DIR, INSTANCE_ID_FILE, vmUuid});
+            vmData.add(new String[]{METATDATA_DIR, VM_ID_FILE, vmUuid});
+        }
+
         vmData.add(new String[]{METATDATA_DIR, PUBLIC_KEYS_FILE, publicKey});
```
Comment (nvazquez) [inline server/src/com/cloud/network/NetworkModelImpl.java]: As `getSshKey(VirtualMachineProfile profile)` method on `ConfigDriveNetworkElement` could return null, shouldn't we check here if it is not null before adding it?

**Comment by DaanHoogland** [comment]
the error I reported earlier, I could not reproduce. ready for merge fwiw

**Pull request merged**